File: clients/java/src/test/java/io/volant/MembershipTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** AddBroker / RemoveBroker / ListMembers client tests against a fake native server. */
class MembershipTest {
    @Test
    void addReturnsGeneration() throws Exception {
        try (MembershipServer srv = new MembershipServer(0, 0, 0)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                long gen = c.addBroker(2, "10.0.0.2", 9092, "r1");
                assertEquals(5L, gen);
            }
            assertEquals(2, srv.addId.get());
            assertEquals("10.0.0.2", srv.addHost.get());
            assertEquals(9092, srv.addPort.get());
            assertEquals("r1", srv.addRack.get());
            assertEquals(Collections.singletonList(Codec.OP_ADD_BROKER), srv.opcodes);
        }
    }

    @Test
    void removeReturnsGeneration() throws Exception {
        try (MembershipServer srv = new MembershipServer(0, 0, 0)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                long gen = c.removeBroker(2);
                assertEquals(6L, gen);
            }
            assertEquals(2, srv.removeId.get());
        }
    }

    @Test
    void listParsesBrokersAndLive() throws Exception {
        try (MembershipServer srv = new MembershipServer(0, 0, 0)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                MembershipList members = c.listMembers();
                assertEquals(4L, members.generation);
                assertEquals(2, members.brokers.size());
                assertEquals(1, members.brokers.get(0).id);
                assertEquals("10.0.0.1", members.brokers.get(0).host);
                assertEquals(9092, members.brokers.get(0).port);
                assertNull(members.brokers.get(0).rack);
                assertEquals(2, members.brokers.get(1).id);
                assertEquals("r1", members.brokers.get(1).rack);
                assertEquals(Arrays.asList(1, 2), members.live);
            }
            assertTrue(srv.listPayloadEmpty);
        }
    }

    @Test
    void addErrorRaises() throws Exception {
        try (MembershipServer srv = new MembershipServer(3, 0, 0)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, () -> c.addBroker(2, "10.0.0.2", 9092));
                assertEquals(3, ex.code);
                assertEquals("add_broker", ex.op);
            }
        }
    }

    @Test
    void removeErrorRaises() throws Exception {
        try (MembershipServer srv = new MembershipServer(0, 2, 0)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, () -> c.removeBroker(2));
                assertEquals(2, ex.code);
                assertEquals("remove_broker", ex.op);
            }
        }
    }

    @Test
    void listErrorRaises() throws Exception {
        try (MembershipServer srv = new MembershipServer(0, 0, 23)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, c::listMembers);
                assertEquals(23, ex.code);
                assertEquals("list_members", ex.op);
            }
        }
    }

    private static final class MembershipServer implements AutoCloseable {
        final int port;
        final AtomicInteger addId = new AtomicInteger();
        final AtomicReference<String> addHost = new AtomicReference<>();
        final AtomicInteger addPort = new AtomicInteger();
        final AtomicReference<String> addRack = new AtomicReference<>();
        final AtomicInteger removeId = new AtomicInteger();
        final List<Integer> opcodes = new CopyOnWriteArrayList<>();
        volatile boolean listPayloadEmpty;
        private final int addError;
        private final int removeError;
        private final int listError;
        private final ServerSocket listen;
        private final Thread thread;

        MembershipServer(int addError, int removeError, int listError) throws IOException {
            this.addError = addError;
            this.removeError = removeError;
            this.listError = listError;
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            thread = new Thread(this::serve, "volant-membership");
            thread.setDaemon(true);
            thread.start();
        }

        private void serve() {
            try (Socket conn = listen.accept()) {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    opcodes.add(d.frame.opcode);
                    byte[] payload;
                    int respOp;
                    if (d.frame.opcode == Codec.OP_ADD_BROKER) {
                        Codec.AddBrokerRequest req = Codec.decodeAddBrokerRequest(d.frame.payload);
                        addId.set(req.id);
                        addHost.set(req.host);
                        addPort.set(req.port);
                        addRack.set(req.rack);
                        payload = Codec.encodeAddBrokerResponse(
                                new Codec.AddBrokerResponse(addError, addError == 0 ? 5L : 0L));
                        respOp = Codec.OP_ADD_BROKER_RESPONSE;
                    } else if (d.frame.opcode == Codec.OP_REMOVE_BROKER) {
                        Codec.RemoveBrokerRequest req = Codec.decodeRemoveBrokerRequest(d.frame.payload);
                        removeId.set(req.id);
                        payload = Codec.encodeRemoveBrokerResponse(
                                new Codec.RemoveBrokerResponse(removeError, removeError == 0 ? 6L : 0L));
                        respOp = Codec.OP_REMOVE_BROKER_RESPONSE;
                    } else if (d.frame.opcode == Codec.OP_LIST_MEMBERS) {
                        listPayloadEmpty = d.frame.payload == null || d.frame.payload.length == 0;
                        payload = Codec.encodeListMembersResponse(
                                new Codec.ListMembersResponse(
                                        listError,
                                        listError == 0 ? 4L : 0L,
                                        Arrays.asList(
                                                new MembershipBroker(1, "10.0.0.1", 9092, null),
                                                new MembershipBroker(2, "10.0.0.2", 9092, "r1")),
                                        Arrays.asList(1, 2)));
                        respOp = Codec.OP_LIST_MEMBERS_RESPONSE;
                    } else {
                        return;
                    }
                    out.write(Frame.encode(respOp, d.frame.correlationId, payload));
                    out.flush();
                }
            } catch (Exception ignored) {
                // test assertions cover the client side
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: clients/go/membership_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type membershipResult struct {
	opcodes     []uint16
	addID       uint32
	addHost     string
	addPort     uint16
	addRack     *string
	removeID    uint32
	listPayload []byte
	err         error
}

func serveMembership(t *testing.T, addErr, removeErr, listErr uint16) (addr string, got *membershipResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &membershipResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			res.opcodes = append(res.opcodes, f.Opcode)
			var (
				payload []byte
				op      uint16
			)
			switch f.Opcode {
			case codec.OpAddBroker:
				req, err := codec.DecodeAddBrokerRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.addID = req.ID
				res.addHost = req.Host
				res.addPort = req.Port
				res.addRack = req.Rack
				gen := uint64(5)
				if addErr != 0 {
					gen = 0
				}
				payload, err = codec.EncodeAddBrokerResponse(codec.AddBrokerResponse{ErrorCode: addErr, Generation: gen})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpAddBrokerResponse
			case codec.OpRemoveBroker:
				req, err := codec.DecodeRemoveBrokerRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.removeID = req.ID
				gen := uint64(6)
				if removeErr != 0 {
					gen = 0
				}
				payload, err = codec.EncodeRemoveBrokerResponse(codec.RemoveBrokerResponse{ErrorCode: removeErr, Generation: gen})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpRemoveBrokerResponse
			case codec.OpListMembers:
				res.listPayload = append([]byte(nil), f.Payload...)
				rack := "r1"
				gen := uint64(4)
				if listErr != 0 {
					gen = 0
				}
				payload, err = codec.EncodeListMembersResponse(codec.ListMembersResponse{
					ErrorCode:  listErr,
					Generation: gen,
					Brokers: []codec.MembershipBroker{
						{ID: 1, Host: "10.0.0.1", Port: 9092, Rack: nil},
						{ID: 2, Host: "10.0.0.2", Port: 9092, Rack: &rack},
					},
					Live: []uint32{1, 2},
				})
				if err != nil {
					res.err = err
					return
				}
				op = codec.OpListMembersResponse
			default:
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
			raw, err := frame.Encode(op, f.CorrelationID, payload)
			if err != nil {
				res.err = err
				return
			}
			if _, err := conn.Write(raw); err != nil {
				res.err = err
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestAddBrokerReturnsGeneration(t *testing.T) {
	addr, got, stop := serveMembership(t, 0, 0, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	rack := "r1"
	gen, err := c.AddBroker(2, "10.0.0.2", 9092, &rack)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if gen != 5 {
		t.Fatalf("generation %d", gen)
	}
	if got.addID != 2 || got.addHost != "10.0.0.2" || got.addPort != 9092 || got.addRack == nil || *got.addRack != "r1" {
		t.Fatalf("add %#v", got)
	}
	if len(got.opcodes) != 1 || got.opcodes[0] != codec.OpAddBroker {
		t.Fatalf("opcodes %#v", got.opcodes)
	}
}

func TestRemoveBrokerReturnsGeneration(t *testing.T) {
	addr, got, stop := serveMembership(t, 0, 0, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	gen, err := c.RemoveBroker(2)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if gen != 6 || got.removeID != 2 {
		t.Fatalf("remove gen=%d id=%d", gen, got.removeID)
	}
}

func TestListMembersParsesBrokersAndLive(t *testing.T) {
	addr, got, stop := serveMembership(t, 0, 0, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	members, err := c.ListMembers()
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if members.Generation != 4 || len(members.Brokers) != 2 || len(members.Live) != 2 {
		t.Fatalf("members %+v", members)
	}
	if members.Brokers[0].ID != 1 || members.Brokers[0].Host != "10.0.0.1" || members.Brokers[0].Rack != nil {
		t.Fatalf("broker0 %+v", members.Brokers[0])
	}
	if members.Brokers[1].ID != 2 || members.Brokers[1].Rack == nil || *members.Brokers[1].Rack != "r1" {
		t.Fatalf("broker1 %+v", members.Brokers[1])
	}
	if members.Live[0] != 1 || members.Live[1] != 2 {
		t.Fatalf("live %+v", members.Live)
	}
	if len(got.listPayload) != 0 {
		t.Fatalf("list request %x", got.listPayload)
	}
}

func TestAddBrokerErrorRaises(t *testing.T) {
	addr, _, stop := serveMembership(t, 3, 0, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.AddBroker(2, "10.0.0.2", 9092, nil)
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 3 || be.Op != "add_broker" {
		t.Fatalf("broker error %+v", be)
	}
}

func TestRemoveBrokerErrorRaises(t *testing.T) {
	addr, _, stop := serveMembership(t, 0, 2, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.RemoveBroker(2)
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "remove_broker" {
		t.Fatalf("broker error %+v", be)
	}
}

func TestListMembersErrorRaises(t *testing.T) {
	addr, _, stop := serveMembership(t, 0, 0, 23)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.ListMembers()
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 23 || be.Op != "list_members" {
		t.Fatalf("broker error %+v", be)
	}
}
